Add the option to do do diffTriplesData using files

Jan-PieterBaert opened the pull request:
This option is configurable with the USE_FILE_DIFF environment variable.
On big datasets this means javascript doesn't have to hold everything in memory and can let `sort` and `comm` do the heavy lifting

cecemel commented on jobs/healing/pipeline-healing.js:
```diff
@@ -229,23 +232,91 @@ async function getScopedSourceTriples( config, property, conceptSchemeUri, publi
+  }
+  return retval;
+}
+function diffFiles(targetFile, sourceFile, S="50%", T="/tmp"){
```
Are we sure new lines are properly escaped always? (if not, thew following example might break it right?)
```
<http://foo>	<http://bar>	""" foo 
bar """ .
```

Jan-PieterBaert replied:
since the files are generated using the `arrayToFile` function and that function uses `JSON.stringify` newlines are properly escaped, e.g. your example becomes `"<http://foo>\t<http://bar>\t\"\"\" foo\nbar \"\"\" ."`

cecemel commented on jobs/healing/pipeline-healing.js:
```diff
@@ -229,23 +232,91 @@ async function getScopedSourceTriples( config, property, conceptSchemeUri, publi
+  let inserts = lines(output1)
+  let deletes = lines(output2)
+
+  execSync(`rm -fv ${sorted1} ${sorted2} ${output1} ${output2}`, optionsNoOutput)
```
is this something that can either be fixed by a lib, or try/catch/finally

Jan-PieterBaert replied:
solved by using lib tmp

cecemel commented on jobs/healing/pipeline-healing.js:
```diff
@@ -229,23 +232,91 @@ async function getScopedSourceTriples( config, property, conceptSchemeUri, publi
+  let deletes = lines(output2)
+
+  execSync(`rm -fv ${sorted1} ${sorted2} ${output1} ${output2}`, optionsNoOutput)
+  execSync(`rm -fv ${targetFile} ${sourceFile}`)
```
same

Jan-PieterBaert replied:
solved by using lib tmp

cecemel commented on jobs/healing/pipeline-healing.js:
```diff
@@ -229,23 +232,91 @@ async function getScopedSourceTriples( config, property, conceptSchemeUri, publi
+  fs.close(fd)
+  return file
+}
+function getTempFile() {
```
perhaps considering a library? (not sure but there is some cleaning boilerplate etc that might be easier)

Jan-PieterBaert replied:
do you have any libraries in mind to do this?

cecemel replied:
lib tmp? (but only if's worth the boilerplate)

Jan-PieterBaert replied:
lib tmp is worth the boilerplate and clearer than before, thanks for the suggestion!

cecemel commented on jobs/healing/pipeline-healing.js:
```diff
@@ -229,23 +232,91 @@ async function getScopedSourceTriples( config, property, conceptSchemeUri, publi
+  let line
+  while ((line = rl.next())) {
+    line = line.toString()
+    if (line[0] === "<" || line[0] === ">") line = line.replace(/^./, "");
```
I don't understand this

Jan-PieterBaert replied:
this was code to strip the first character away for the output of the `diff` command, however it's been rendered useless since the `comm` command is used now

cecemel commented on jobs/healing/pipeline-healing.js:
```diff
@@ -229,23 +232,91 @@ async function getScopedSourceTriples( config, property, conceptSchemeUri, publi
+  while ((line = rl.next())) {
+    line = line.toString()
+    if (line[0] === "<" || line[0] === ">") line = line.replace(/^./, "");
+    retval.push(JSON.parse(line) || {})
```
neither the json parse nor the || 

Comments might be welcome :-)

Jan-PieterBaert replied:
the json parse is to convert the line back to an Object, the || was put there to avoid parse errors stopping the entire pipeline.
I will add a comment for it!

cecemel replied:
but not sure how line can be valid json string, and how || will catch parse exceptions? I miss somthing.

Jan-PieterBaert replied:
line is a valid json string because the dataset gets converted into lines by the `arrayToFile` function which makes each element a single line (using JSON.stringify) which means here we parse such line which comes from the stringify function

cecemel commented:
@Jan-PieterBaert thanks, also perhaps give it a tag, so we can test in on a huge deployed QA system.

Jan-PieterBaert commented:
@cecemel I have solved your remarks and made this PR available on dockerhub with the tag `janpieterbaert/delta-producer-publication-graph-maintainer:feature-delta-using-files`

cecemel commented:
> janpieterbaert/delta-producer-publication-graph-maintainer:feature-delta-using-files

all right let's roll :sunglasses:

cecemel commented:
@Jan-PieterBaert 
it seems throwing errors on the lblod-harvester
```
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494684616Z <ref *1> Error: spawnSync /bin/sh ENOBUFS
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494704657Z     at Object.spawnSync (internal/child_process.js:1067:20)
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494711059Z     at spawnSync (child_process.js:602:24)
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494715979Z     at execSync (child_process.js:647:15)
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494720769Z     at diffFiles (/usr/src/output/app/jobs/healing/pipeline-healing.js:285:31)
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494725578Z     at diffTriplesData (/usr/src/output/app/jobs/healing/pipeline-healing.js:315:12)
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494730358Z     at getSourceTriples (/usr/src/output/app/jobs/healing/pipeline-healing.js:152:19)
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494735077Z     at runMicrotasks (<anonymous>)
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494739867Z     at processTicksAndRejections (internal/process/task_queues.js:93:5)
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494744777Z     at async runHealingTask (/usr/src/output/app/jobs/healing/pipeline-healing.js:63:29)
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494749546Z     at async executeHealingTask (/usr/src/output/app/jobs/healing/main.js:39:17) {
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494754216Z   errno: -105,
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494758705Z   code: 'ENOBUFS',
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494763193Z   syscall: 'spawnSync /bin/sh',
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494767713Z   path: '/bin/sh',
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494772181Z   spawnargs: [
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494776630Z     '-c',
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494781360Z     'comm -23 /tmp/tmp-1-ywMboVtn41qF /tmp/tmp-1-iPm9izZGpLzt | tee /tmp/tmp-1-QRyk2sCA1snV'
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494786139Z   ],
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494790648Z   error: [Circular *1],
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494795157Z   status: null,
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494799616Z   signal: 'SIGTERM',
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494804135Z   output: [
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494808564Z     null,
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494813344Z     <Buffer 7b 22 6e 54 72 69 70 6c 65 22 3a 22 3c 68 74 74 70 3a 2f 2f 64 61 74 61 2e 6c 62 6c 6f 64 2e 69 6e 66 6f 2f 61 72 74 69 63 6c 65 73 2f 30 35 62 61 61 ... 1054276 more bytes>,
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494818955Z     <Buffer >
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494823644Z   ],
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494828083Z   pid: 28,
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494832682Z   stdout: <Buffer 7b 22 6e 54 72 69 70 6c 65 22 3a 22 3c 68 74 74 70 3a 2f 2f 64 61 74 61 2e 6c 62 6c 6f 64 2e 69 6e 66 6f 2f 61 72 74 69 63 6c 65 73 2f 30 35 62 61 61 ... 1054276 more bytes>,
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494861590Z   stderr: <Buffer >
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.494868223Z }
delta-producer-publication-graph-maintainer-besluiten_1  | 2023-05-04T09:23:49.495088554Z default-processing-queue: Error while processing delta in queue TypeError: value.replace is not a function
```
full logs in DM

Jan-PieterBaert commented:
@cecemel this buffer error was probably caused because `execSync` tried to store the output while it is not necessay, I've updated the options for that command, so this error shouldn't happen anymore
I have also updated the `janpieterbaert/delta-producer-publication-graph-maintainer:feature-delta-using-files` container on dockerhub

Jan-PieterBaert commented:
replaced by #16
